Sources_v3/Controllers/CurrentUserController.swift

```diff
@@ -0,0 +1,287 @@
+    private lazy var currentUserObserver = createUserObserver()
+        .onChange { [unowned self] change in
+            self.delegateCallback {
+                $0?.currentUserController(self, didChangeCurrentUser: change)
+            }
+        }
+        .onFieldChange(\.unreadCount) { [unowned self] change in
+            self.delegateCallback {
+                $0?.currentUserController(self, didChangeCurrentUserUnreadCount: change.unreadCount)
+            }
+        }
+
+    /// A type-erased delegate.
+    private(set) var anyDelegate: AnyCurrentUserControllerDelegate<ExtraData>? {
+        didSet { stateDelegate = anyDelegate }
+    }
+    
+    /// The currently logged-in user.
+    /// Always returns `nil` if `startUpdating` was not called
+    /// To observe the updates of this value, set your class as a delegate of this controller and call `startUpdating`.
+    public var currentUser: CurrentUserModel<ExtraData.User>? {
+        guard state != .inactive else {
+            log.warning("Accessing `currentUser` fields before calling `startUpdating()` always results in `nil`.")
+            return nil
+        }
+
+        return currentUserObserver.item
+    }
+
+    /// The unread messages and channels count for the current user.
+    /// Always returns `noUnread` if `startUpdating` was not called.
+    /// To observe the updates of this value, set your class as a delegate of this controller and call `startUpdating`.
+    public var unreadCount: UnreadCount {
+        currentUser?.unreadCount ?? .noUnread
+    }
+
+    /// Creates a new `CurrentUserControllerGeneric`.
+    /// - Parameters:
+    ///   - client: The `Client` instance this controller belongs to.
+    ///   - environment: The source of internal dependencies
+    init(client: Client<ExtraData>, environment: Environment<ExtraData>) {
+        self.client = client
+        self.environment = environment
+    }
+    
+    /// Starts updating the results.
+    ///
+    /// It **synchronously** loads the data for the referenced objects from the local cache.
+    /// The `currentUser` and `unreadCount` properties are immediately available once this method returns.
+    /// Any further changes to the data are communicated using `delegate`.
+    ///
+    /// - Parameter completion: Called when the controller has finished fetching data from a database.
+    /// If the data fetching fails, the `error` variable contains more details about the problem.
+    public func startUpdating(_ completion: ((Error?) -> Void)? = nil) {
+        do {
+            try currentUserObserver.startObserving()
+        } catch {
+            callback { completion?(ClientError.FetchFailed()) }
+            return
+        }
+        
+        state = .localDataFetched
+
+        callback { completion?(nil) }
+    }
+}
+
+// MARK: - Environment
+
+extension CurrentUserControllerGeneric {
+    struct Environment<ExtraData: ExtraDataTypes> {
```

Comment:
I think you can remove the `<ExtraData: ExtraDataTypes>` from Environment. It automatically captures the extra data from the wrapping `CurrentUserControllerGeneric `

Reply:
fixed in 5fa443a574cff4cdfb1272bc1b0b7d13afcfc6d0